src/NetSuiteClient.php

```diff
@@ -73,19 +78,91 @@ public function setDataCenterUrl(array $config)
         $this->client->__setLocation($dataCenterUrl);
     }
 
-    public static function createFromEnv($options = array(), $client = null)
+    /**
+     * Create a configuration array by inspecting the $_ENV superglobal.
+     *
+     * @return array
+     */
+    public static function getEnvConfig()
     {
-        $config = array(
-            'endpoint' => getenv('NETSUITE_ENDPOINT') ?: '2017_1',
-            'host' => getenv('NETSUITE_HOST') ?: 'https://webservices.sandbox.netsuite.com',
-            'email' => getenv('NETSUITE_EMAIL'),
-            'password' => getenv('NETSUITE_PASSWORD'),
-            'role' => getenv('NETSUITE_ROLE') ?: '3',
-            'account' => getenv('NETSUITE_ACCOUNT'),
-            'app_id' => getenv('NETSUITE_APP_ID') ?: '4AD027CA-88B3-46EC-9D3E-41C6E6A325E2',
-            'logging' => getenv('NETSUITE_LOGGING'),
-            'log_path' => getenv('NETSUITE_LOG_PATH'),
-        );
+        $config = [
+            'endpoint'           => getenv('NETSUITE_ENDPOINT') ?: '2019_1',
+            'host'               => getenv('NETSUITE_HOST') ?: 'https://webservices.sandbox.netsuite.com',
+            'email'              => getenv('NETSUITE_EMAIL'),
+            'password'           => getenv('NETSUITE_PASSWORD'),
+            'role'               => getenv('NETSUITE_ROLE') ?: '3',
+            'account'            => getenv('NETSUITE_ACCOUNT'),
+            'app_id'             => getenv('NETSUITE_APP_ID') ?: '4AD027CA-88B3-46EC-9D3E-41C6E6A325E2',
+            'logging'            => getenv('NETSUITE_LOGGING'),
+            'log_path'           => getenv('NETSUITE_LOG_PATH'),
+        ];
+
+        // These config keys aren't required by all users, but if they are
+        // defined in the config array, then they must be correct, thus we
+        // will omit ones that have been left empty in the .env file.
+        $optKeys = [
+            'NETSUITE_CONSUMER_KEY'    => 'consumerKey',
+            'NETSUITE_CONSUMER_SECRET' => 'consumerSecret',
+            'NETSUITE_TOKEN_KEY'       => 'token',
+            'NETSUITE_TOKEN_SECRET'    => 'tokenSecret',
+            'NETSUITE_HASH_TYPE'       => 'signatureAlgorithm'
+        ];
+        foreach ($optKeys as $optKey => $cfgKey) {
+            if ($optVal = getenv($optKey)) {
+                $config[$cfgKey] = $optVal;
+            }
+        }
+
+        return $config;
+    }
+
+    /**
+     * Make sure that this client object has at least the basic required
+     * configuration values defined or else throw a runtime exception.
+     *
+     * @param array $config
+     *
+     * @return void
+     */
+    public function validateConfig(array $config)
+    {
+        $requiredParams = [
+            'endpoint',
+            'host',
+            'email',
```

Comment:
Given username/password are not required for token-based authentication, won't this return a false negative if the user is using TBA?

Reply:
Assuming of course that they've chosen to omit those redundant values from their config.

Reply:
Thanks for the feedback! Because I only work with one type of authentication, I was not entirely sure about which fields exactly must be required for each of the auth methods. As I think it's preferable to avoid infusing much logic specific to NetSuite into this library, I'll make an adjustment to this method so that we only require what is truly required for all connections and if the user fails to provide an appropriate combination of authentication keys (like username and password or token-based values) then the resulting error which will pass through will make this easy enough to correct.

Reply:
Changes look good